src/gainmap.c

```diff
@@ -49,4 +53,238 @@ avifBool avifGainMapMetadataFractionsToDouble(avifGainMapMetadataDouble * dst, c
     return AVIF_TRUE;
 }
 
+// ---------------------------------------------------------------------------
+
+// Returns a weight in [-1.0, 1.0] that represents how much the gain map should be applied.
+static float avifGetGainMapWeight(float hdrCapacity, const avifGainMapMetadataDouble * metadata)
+{
+    const float hdrCapacityMin = (float)metadata->hdrCapacityMin;
+    const float hdrCapacityMax = (float)metadata->hdrCapacityMax;
+    float w = 0.0f;
+    if (hdrCapacity > hdrCapacityMin) {
+        if (hdrCapacity < hdrCapacityMax) {
+            w = (logf(hdrCapacity) - logf(hdrCapacityMin)) / (logf(hdrCapacityMax) - logf(hdrCapacityMin));
+        } else {
+            w = 1.0f;
+        }
+    }
+
+    if (metadata->baseRenditionIsHDR) {
+        w -= 1.0f;
+    }
+
+    return w;
+}
+
+// Linear interpolation between 'a' and 'b' (returns 'a' if w == 0.0f, returns 'b' if w == 1.0f).
+static inline float lerp(float a, float b, float w)
+{
+    return (1.0f - w) * a + w * b;
+}
+
+#define SDR_WHITE_NITS 203.0f
+
+avifResult avifImageApplyGainMapRGB(const avifRGBImage * baseImage,
+                                    avifTransferCharacteristics transferCharacteristics,
+                                    const avifGainMap * gainMap,
+                                    float hdrCapacity,
+                                    avifTransferCharacteristics outputTransferCharacteristics,
+                                    avifRGBImage * toneMappedImage,
+                                    avifContentLightLevelInformationBox * clli,
+                                    avifDiagnostics * diag)
```

Comment:
It's definitely simpler, but then we won't get the benefit of the diagonistics? I think the avifResult return status alone is difficult to debug... Right now avifDiagnostics is a bit annoying because it cannot be null, but I think we should also revisit this.

I saw that in [this comment](https://github.com/AOMediaCodec/libavif/pull/937#discussion_r1347560842) you said that it had to be non-null currently because "it's a useful debugging tool", but then you don't seem to think its use is warranted here or in a lot of other places.
I think making it nullable would allow the caller to decide if they want the extra diagnostics or not?

Reply:
The main point of the `diag` parameter is to give a detailed explanation when parsing something really complicated. If the only way a function can fail is bad arguments or memory allocation failures, then it probably doesn't need the `diag` parameter. That's why I think the `diag` parameter should only be used with functions that deal with the ISO BMFF file format. We can keep the `diag` parameter if it is useful to these two functions. (I didn't analyze the diagnostics messages written by these two functions.)

We can allow `diag` to be null. But I am worried that we may see calls to those API functions that end with a `/*diag=*/NULL` argument, which seems like an eyesore. That's what I meant by "the functions will be simpler without the `diag` parameter".

Reply:
Indeed apart form memory issues, errors should come from bad inputs. But even then, there are many ways in which the inputs can be wrong. I find it very frustrating to debug a simple status return value.